cluster_experiments/washover.py

```diff
@@ -242,5 +242,124 @@ def from_config(cls, config) -> "Washover":
         return cls(washover_time_delta=washover_time_delta)
 
 
+class TwoEventsWashover(Washover):
+    """
+    For each record in our experiment, two events' timestamps,
+    login_timestamp and logout_timestamp.
+     We want to apply washover such that if there is a change in treatment
+    between login_timestamp and logout_timestamp.
+    Ex:
+    Calendar:
+        +-----------+-------+
+        | treatment | time  |
+        +-----------+-------+
+        | A         | 10:00 |
+        | B         | 11:00 |
+        | B         | 12:00 |
+        +-----------+-------+
+    Record data:
+
+        +----+------------+----------+
+        | id | start_time | end_time |
+        +----+------------+----------+
+        |  1 | 10:50      | 10:59    |
+        |  2 | 10:51      | 11:01    |
+        |  3 | 11:01      | 11:05    |
+        |  4 | 11:01      | 12:01    |
+        |  5 | 12:01      | 12:05    |
+        +----+------------+----------+
+
+    This washover will drop record number 2.
+    """
+
+    def __init__(
+        self,
+        calendar_df: pd.DataFrame,
+        time_column_calendar: str = "start_time",
+        treatment_column_calendar: str = "treatment",
+    ):
+        """
+        Args:
+            calendar_df (pd.DataFrame): df with the calendar of treatments.
+            time_column_calendar (str): Name of the time column in the calendar df.
+            treatment_column_calendar (str): Name of the treatment column in the calendar df.
+            record_df (pd.DataFrame): Record dataframe.
+            start_time_column (str): Name of the start time column in the record dataframe.
+            end_time_column (str): Name of the end time column in the record dataframe.
+
+        """
+        self.calendar_df = calendar_df
+        self.time_column_calendar = time_column_calendar
+        self.treatment_column_calendar = treatment_column_calendar
+
+    def _validate_columns(
+        self, record_df: pd.DataFrame, start_time_column: str, end_time_column: str
+    ):
+        if self.time_column_calendar not in self.calendar_df.columns:
+            raise ValueError(
+                f"{self.time_column_calendar = } is not in the calendar dataframe columns and/or not specified as an input."
+            )
+        if self.treatment_column_calendar not in self.calendar_df.columns:
+            raise ValueError(
+                f"{self.treatment_column_calendar = } is not in the calendar dataframe columns and/or not specified as an input."
+            )
+        if start_time_column not in record_df.columns:
+            raise ValueError(
+                f"{start_time_column = } is not in the record dataframe columns and/or not specified as an input."
+            )
+        if end_time_column not in record_df.columns:
+            raise ValueError(
+                f"{end_time_column = } is not in the record dataframe columns and/or not specified as an input."
+            )
+
+    def washover(
```

Comment:
I'd try to maintain the signature of the abstract washover class, if not it's very hard to put it into power analysis:
```python
    @abstractmethod
    def washover(
        self,
        df: pd.DataFrame,
        truncated_time_col: str,
        treatment_col: str,
        cluster_cols: List[str],
        original_time_col: Optional[str] = None,
    ) -> pd.DataFrame:
        """Abstract method to add washvover to the dataframe."""
```

Reply:
For this reason, I'd send start and end columns to the init method